// graphql-tools-fork/dist/wrap/transforms/index.js
Object.defineProperty(exports, "__esModule", { value: true });
var CheckResultAndHandleErrors_1 = require("./CheckResultAndHandleErrors");
exports.CheckResultAndHandleErrors = CheckResultAndHandleErrors_1.default;
var ExpandAbstractTypes_1 = require("./ExpandAbstractTypes");
exports.ExpandAbstractTypes = ExpandAbstractTypes_1.default;
var AddReplacementSelectionSets_1 = require("./AddReplacementSelectionSets");
exports.AddReplacementSelectionSets = AddReplacementSelectionSets_1.default;
var AddMergedTypeSelectionSets_1 = require("./AddMergedTypeSelectionSets");
exports.AddMergedTypeSelectionSets = AddMergedTypeSelectionSets_1.default;
var AddArgumentsAsVariables_1 = require("./AddArgumentsAsVariables");
exports.AddArgumentsAsVariables = AddArgumentsAsVariables_1.default;
var FilterToSchema_1 = require("./FilterToSchema");
exports.FilterToSchema = FilterToSchema_1.default;
var AddTypenameToAbstract_1 = require("./AddTypenameToAbstract");
exports.AddTypenameToAbstract = AddTypenameToAbstract_1.default;
var RenameTypes_1 = require("./RenameTypes");
exports.RenameTypes = RenameTypes_1.default;
var FilterTypes_1 = require("./FilterTypes");
exports.FilterTypes = FilterTypes_1.default;
var RenameRootTypes_1 = require("./RenameRootTypes");
exports.RenameRootTypes = RenameRootTypes_1.default;
var TransformCompositeFields_1 = require("./TransformCompositeFields");
exports.TransformCompositeFields = TransformCompositeFields_1.default;
var TransformRootFields_1 = require("./TransformRootFields");
exports.TransformRootFields = TransformRootFields_1.default;
var RenameRootFields_1 = require("./RenameRootFields");
exports.RenameRootFields = RenameRootFields_1.default;
var FilterRootFields_1 = require("./FilterRootFields");
exports.FilterRootFields = FilterRootFields_1.default;
var TransformObjectFields_1 = require("./TransformObjectFields");
exports.TransformObjectFields = TransformObjectFields_1.default;
var RenameObjectFields_1 = require("./RenameObjectFields");
exports.RenameObjectFields = RenameObjectFields_1.default;
var FilterObjectFields_1 = require("./FilterObjectFields");
exports.FilterObjectFields = FilterObjectFields_1.default;
var TransformInterfaceFields_1 = require("./TransformInterfaceFields");
exports.TransformInterfaceFields = TransformInterfaceFields_1.default;
var RenameInterfaceFields_1 = require("./RenameInterfaceFields");
exports.RenameInterfaceFields = RenameInterfaceFields_1.default;
var FilterInterfaceFields_1 = require("./FilterInterfaceFields");
exports.FilterInterfaceFields = FilterInterfaceFields_1.default;
var TransformQuery_1 = require("./TransformQuery");
exports.TransformQuery = TransformQuery_1.default;
var ExtendSchema_1 = require("./ExtendSchema");
exports.ExtendSchema = ExtendSchema_1.default;
var WrapType_1 = require("./WrapType");
exports.WrapType = WrapType_1.default;
var WrapFields_1 = require("./WrapFields");
exports.WrapFields = WrapFields_1.default;
var HoistField_1 = require("./HoistField");
exports.HoistField = HoistField_1.default;
var MapFields_1 = require("./MapFields");
exports.MapFields = MapFields_1.default;
// superseded by AddReplacementFragments
var ReplaceFieldWithFragment_1 = require("./ReplaceFieldWithFragment");
exports.ReplaceFieldWithFragment = ReplaceFieldWithFragment_1.default;
// superseded by AddReplacementSelectionSets
var AddReplacementFragments_1 = require("./AddReplacementFragments");
exports.AddReplacementFragments = AddReplacementFragments_1.default;
// superseded by TransformQuery
var WrapQuery_1 = require("./WrapQuery");
exports.WrapQuery = WrapQuery_1.default;
var ExtractField_1 = require("./ExtractField");
exports.ExtractField = ExtractField_1.default;
