// graphql-tools-fork/dist/wrap/transforms/RenameTypes.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var index_1 = require("../../polyfills/index");
var Interfaces_1 = require("../../Interfaces");
var index_2 = require("../../utils/index");
var RenameTypes = /** @class */ (function () {
    function RenameTypes(renamer, options) {
        this.renamer = renamer;
        this.map = {};
        this.reverseMap = {};
        var _a = options != null ? options : {}, _b = _a.renameBuiltins, renameBuiltins = _b === void 0 ? false : _b, _c = _a.renameScalars, renameScalars = _c === void 0 ? true : _c;
        this.renameBuiltins = renameBuiltins;
        this.renameScalars = renameScalars;
    }
    RenameTypes.prototype.transformSchema = function (originalSchema) {
        var _a;
        var _this = this;
        return index_2.mapSchema(originalSchema, (_a = {},
            _a[Interfaces_1.MapperKind.TYPE] = function (type) {
                if (index_1.isSpecifiedScalarType(type) && !_this.renameBuiltins) {
                    return undefined;
                }
                if (graphql_1.isScalarType(type) && !_this.renameScalars) {
                    return undefined;
                }
                var oldName = type.name;
                var newName = _this.renamer(oldName);
                if (newName && newName !== oldName) {
                    _this.map[oldName] = type.name;
                    _this.reverseMap[newName] = oldName;
                    var newConfig = __assign(__assign({}, index_1.toConfig(type)), { name: newName });
                    if (graphql_1.isObjectType(type)) {
                        return new graphql_1.GraphQLObjectType(newConfig);
                    }
                    else if (graphql_1.isInterfaceType(type)) {
                        return new graphql_1.GraphQLInterfaceType(newConfig);
                    }
                    else if (graphql_1.isUnionType(type)) {
                        return new graphql_1.GraphQLUnionType(newConfig);
                    }
                    else if (graphql_1.isInputObjectType(type)) {
                        return new graphql_1.GraphQLInputObjectType(newConfig);
                    }
                    else if (graphql_1.isEnumType(type)) {
                        return new graphql_1.GraphQLEnumType(newConfig);
                    }
                    else if (graphql_1.isScalarType(type)) {
                        return new graphql_1.GraphQLScalarType(newConfig);
                    }
                    throw new Error("Unknown type " + type + ".");
                }
            },
            _a[Interfaces_1.MapperKind.ROOT_OBJECT] = function () {
                return undefined;
            },
            _a));
    };
    RenameTypes.prototype.transformRequest = function (originalRequest) {
        var _a;
        var _this = this;
        var newDocument = graphql_1.visit(originalRequest.document, (_a = {},
            _a[graphql_1.Kind.NAMED_TYPE] = function (node) {
                var name = node.name.value;
                if (name in _this.reverseMap) {
                    return __assign(__assign({}, node), { name: {
                            kind: graphql_1.Kind.NAME,
                            value: _this.reverseMap[name],
                        } });
                }
            },
            _a));
        return {
            document: newDocument,
            variables: originalRequest.variables,
        };
    };
    RenameTypes.prototype.transformResult = function (result) {
        return __assign(__assign({}, result), { data: this.renameTypes(result.data) });
    };
    RenameTypes.prototype.renameTypes = function (value) {
        var _this = this;
        if (value == null) {
            return value;
        }
        else if (Array.isArray(value)) {
            value.forEach(function (v, index) {
                value[index] = _this.renameTypes(v);
            });
            return value;
        }
        else if (typeof value === 'object') {
            Object.keys(value).forEach(function (key) {
                value[key] =
                    key === '__typename'
                        ? _this.renamer(value[key])
                        : _this.renameTypes(value[key]);
            });
            return value;
        }
        return value;
    };
    return RenameTypes;
}());
exports.default = RenameTypes;
